adapters/Vungle/VungleAdapter/Bidding/GADMediationVungleNativeAd.m

```diff
@@ -0,0 +1,248 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "GADMediationVungleNativeAd.h"
+#include <stdatomic.h>
+#import "GADMAdapterVungleUtils.h"
+#import "GADMAdapterVungleRouter.h"
+#import <VungleSDK/VungleNativeAd.h>
+
+@interface GADMediationVungleNativeAd () <GADMediationNativeAd, VungleNativeAdDelegate, GADMAdapterVungleDelegate>
+
+@end
+
+@implementation GADMediationVungleNativeAd {
+  /// Ad configuration for the ad to be loaded.
+  GADMediationNativeAdConfiguration *_adConfiguration;
+
+  /// The completion handler to call when the ad loading succeeds or fails.
+  GADMediationNativeLoadCompletionHandler _adLoadCompletionHandler;
+
+  /// The Vungle native ad.
+  VungleNativeAd *_nativeAd;
+
+  /// The ad event delegate to forward ad rendering events to the Google Mobile Ads SDK.
+  id<GADMediationNativeAdEventDelegate> _delegate;
+    
+  /// The Vungle container for the main ad
+  VungleMediaView *_mediaView;
+}
+
+@synthesize desiredPlacement;
+@synthesize isAdLoaded;
+
+- (nonnull instancetype)initNativeAdForAdConfiguration:(nonnull GADMediationNativeAdConfiguration *)adConfiguration
+                                     completionHandler:(nonnull GADMediationNativeLoadCompletionHandler)completionHandler {
+  self = [super init];
+  if (self) {
+    _adConfiguration = adConfiguration;
+
+    // Store the ad config and completion handler for later use.
+    __block atomic_flag adLoadHandlerCalled = ATOMIC_FLAG_INIT;
+    __block GADMediationNativeLoadCompletionHandler originalCompletionHandler = [completionHandler copy];
+
+    // Ensure the original completion handler is only called once, and is deallocated once called.
+    _adLoadCompletionHandler = ^id<GADMediationNativeAdEventDelegate>(_Nullable id<GADMediationNativeAd> ad, NSError *_Nullable error) {
+      if (atomic_flag_test_and_set(&adLoadHandlerCalled)) {
+        return nil;
+      }
+      id<GADMediationNativeAdEventDelegate> delegate = nil;
+      if (originalCompletionHandler) {
+        delegate = originalCompletionHandler(ad, error);
+      }
+      originalCompletionHandler = nil;
+      return delegate;
+    };
+  }
+  return self;
+}
+
+- (void)requestAd {
```

Comment:
Nit, `requestNativeAd`.

Reply:
I will address this in another commit. It is in internal review currently.